src/socket/socket_ref.rs

```diff
@@ -0,0 +1,83 @@
+use core::ops::{Deref, DerefMut};
+use socket::{FromSocket, RawSocket, Socket, TcpSocket, UdpSocket, SocketHandle};
+
+/// A trait for tracking a socket usage sessions.
+///
+/// Allows implementation of custom on-drop logic.
+pub trait SocketSession {
+    fn finish(&mut self) {}
+}
+
+impl<'a, 'b> SocketSession for RawSocket<'a, 'b> {}
+impl<'a, 'b> SocketSession for UdpSocket<'a, 'b> {}
+impl<'a> SocketSession for TcpSocket<'a> {}
+
+impl<'a, 'b> SocketSession for Socket<'a, 'b> {
+    fn finish(&mut self) {
+        match *self {
+            Socket::Raw(ref mut raw_socket) => {
+                raw_socket.finish();
+            }
+            Socket::Udp(ref mut udp_socket) => {
+                udp_socket.finish();
+            }
+            Socket::Tcp(ref mut tcp_socket) => {
+                tcp_socket.finish();
+            }
+            Socket::__Nonexhaustive => unreachable!(),
+        }
+    }
+}
+
+/// A tracking smart-pointer to a socket.
+///
+/// Implements `Deref` and `DerefMut` to the socket it contains.
+pub struct SocketRef<'a, T: SocketSession + 'a> {
+    socket: &'a mut T,
+    handle: SocketHandle,
+    fused: bool,
+}
+
+impl<'a, T: SocketSession> SocketRef<'a, T> {
+    pub(crate) fn new(socket: &'a mut T, handle: SocketHandle) -> Self {
+        SocketRef {
+            socket,
+            handle,
+            fused: true,
+        }
+    }
+}
+
+impl<'a, 'b, 'c, T: FromSocket<'b, 'c>> SocketRef<'a, T> {
+    pub(crate) fn from_socket_ref(mut socket_ref: SocketRef<'a, Socket<'b, 'c>>)
+                                  -> Option<SocketRef<'a, T>> {
+        if let Some(socket) = T::from_socket(socket_ref.socket) {
+            socket_ref.fused = false;
```

Comment:
It's possible I'm missing something here: I want to prevent a spurious destructor call, but I can't use `mem::forget` since `socket_ref` is being borrowed from until the end of the method. It doesn't prevent the destructor from being called tho, so I have to do it by hand.

Reply:
`Option<&mut T>` maybe? If I understand correctly, this Option would be deconstructed exactly three times, one in constructor, one in destructor and one in deref, so the one added jump is basically nothing. Then just take the option.

Not worth introducing unsafe code just for this IMO.

Reply:
`Option<&mut T>` in `SocketRef` would mean `DerefMut::deref_mut` could panic, and there's really no reason for it. I guess I live it as it is for now.